### ISSUE TITLE
Stable sorting of tables

### ISSUE THREAD

**Issue by schanzer** [brownplt/pyret-lang#1695]
Bootstrap:DS switched from method syntax to function syntax, and we're already seeing huge gains from teachers. Coupled with the Circles of Evaluation, they are _immediately_ more comfortable composing operations. One of the most common cases is composing sorting operations, for example "sort animals alphabetically by species, then by age". 

Our `order-by` method makes no guarantees about stability, of course, so teachers have been consistently surprised by the results! 

What's fascinating is that this hasn't been an issue for the past 7 years with method syntax -- even though it's the exact same behavior. That suggests that methods were either hard enough that teachers weren't composing them, or that they were mysterious enough that teachers didn't have any confidence their own expectations of how they should work.

Anyway, this has now bubbled up to the point where it's become a sticking point, so I'm wondering what our options are here. Pyret appears to walk the table from the bottom up when sorting, presumably for performance/memory reasons. Is that consistent enough that I could invert the process in the function that wraps the call to `.order-by`? Or would this require a change in the runtime itself?

**Comment by shriram** [comment]
Thanks. Please tag Luke West there.

**Comment by schanzer** [comment]
Major kudos to @blerner for jumping in with a quick patch. It's now live on the web! I'll remove it when this issue is closed.

**Comment by schanzer** [comment]
Tagging @westluke as per Shriram's request

**Comment by blerner** [comment]
No motion on this in a year, so I wrote up a merge-sort implementation for lists.  Note that I did _not_ replace the existing implementation of `sort` and `sort-by`, but instead added `stable-sort` and `stable-sort-by` (in both method and function form).  I then revised tables to use `stable-sort-by` instead.  This seems to me the minimally-invasive approach that doesn't break any existing behavior.  If we *choose* to change the stability behavior of `sort` itself, that is a breaking change and we should decide on it *soon* while we're first creating `essentials2024`...

(BTW: Also note, @jpolitz  and @shriram , that it's _really ugly_ to write an imperative while-loop-based algorithm in Pyret :(  I got as close as I could to a named-let (let loop ...) construct, and I kinda hate it (even more than I dislike named-let ;-) )  I could not think of any way to kludge a Pyret `for` loop into a syntactically nice while-loop.)
